core/embed/trezorhal/unix/touch/touch.c

```diff
@@ -31,17 +31,145 @@
 extern int sdl_display_res_x, sdl_display_res_y;
 extern int sdl_touch_offset_x, sdl_touch_offset_y;
 
-static bool _touch_detected = false;
 static int _touch_x = 0;
 static int _touch_y = 0;
 
+// distance from the edge where arrow button swipe starts [px]
+static const int _btn_swipe_begin = 20;
+// length of the arrow button swipe [px]
+static const int _btn_swipe_length = 60;
+
+// A state machine to handle both mouse inputs (simulating touch) and arrow
+// buttons (to simulate scroll movements). The variable `input_state` is used to
+// ensure that arrow keys are not processed when mouse input is in progress and
+// that mouse actions are not processed while arrow button swipe is not
+// finished.
+typedef enum {
+  IDLE,
+  MOUSE_ACTIVE,
+  MOUSE_OUTSIDE,
+  BUTTON_SWIPE_INITIATED,
+  BUTTON_SWIPE_COMPLETED
+} TouchInputState;
+
+static TouchInputState input_state = IDLE;
+
 static bool is_inside_display(int x, int y) {
-  return (x >= sdl_touch_offset_x && y >= sdl_touch_offset_y &&
-          x - sdl_touch_offset_x < sdl_display_res_x &&
-          y - sdl_touch_offset_y < sdl_display_res_y);
+  return x >= sdl_touch_offset_x && y >= sdl_touch_offset_y &&
+         x - sdl_touch_offset_x < sdl_display_res_x &&
+         y - sdl_touch_offset_y < sdl_display_res_y;
+}
+
+void handle_mouse_events(SDL_Event event, int* ev_type, int* ev_x, int* ev_y) {
+  bool inside_display = is_inside_display(event.button.x, event.button.y);
+  switch (event.type) {
+    case SDL_MOUSEBUTTONDOWN:
+      if (inside_display) {
+        *ev_x = event.button.x - sdl_touch_offset_x;
+        *ev_y = event.button.y - sdl_touch_offset_y;
+        *ev_type = TOUCH_START;
+        input_state = MOUSE_ACTIVE;
+      }
+      break;
+
+    case SDL_MOUSEBUTTONUP:
+      if (input_state != IDLE) {
+        *ev_x = inside_display ? event.button.x - sdl_touch_offset_x : _touch_x;
+        *ev_y = inside_display ? event.button.y - sdl_touch_offset_y : _touch_y;
+        *ev_type = TOUCH_END;
+        input_state = IDLE;
+      }
+      break;
+
+    case SDL_MOUSEMOTION:
+      if (input_state != IDLE) {
+        if (inside_display) {
+          *ev_x = event.motion.x - sdl_touch_offset_x;
+          *ev_y = event.motion.y - sdl_touch_offset_y;
+          // simulate TOUCH_START if pressed in mouse returned on visible area
+          *ev_type = (input_state == MOUSE_OUTSIDE) ? TOUCH_START : TOUCH_MOVE;
+          input_state = MOUSE_ACTIVE;
+        } else {
+          // use last valid coordinates and simulate TOUCH_END
```

Comment:
If you leave the window with the mouse button down and continue moving, this code will be executed and it will generate a continuous stream of TOUCH_END events.

Reply:
Good catch. Now the TOUCH_END is set only once per leaving.
https://github.com/trezor/trezor-firmware/pull/3619/commits/f248d6218433721716ffe2372555eb65fdeb0fc2